File: source/python/13. Roman to Integer.py
ROMAN = {
    'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000
}


class Solution:
    def romanToInt(self, s: str) -> int:
        result = 0
        for idx, char in enumerate(s):
            if idx < len(s) - 1 and ROMAN[char] < ROMAN[s[idx + 1]]:
                result -= ROMAN[char]
            else:
                result += ROMAN[char]

        return result


def test():
    solution = Solution()
    test_cases = [
        ("III", 3),
        ("IV", 4),
        ("IX", 9),
        ("LVIII", 58),
        ("MCMXCIV", 1994),
    ]
    for s, expected in test_cases:
        assert solution.romanToInt(
            s) == expected, f"{s=}, {expected=}, {solution.romanToInt(s)=}"

    print("All tests passed!")


if __name__ == "__main__":
    test()

File: source/cpp/13. Roman to Integer.cpp
#include <string_view>
#include <unordered_map>


const std::unordered_map<char, uint16_t> roman_map = {
    {'I', 1},
    {'V', 5},
    {'X', 10},
    {'L', 50},
    {'C', 100},
    {'D', 500},
    {'M', 1000},
};

class Solution {
public:
    uint16_t romanToInt(std::string_view s) {
        uint16_t result = 0;

        for (auto i = 0; i < s.size(); ++i) {
            if (i + 1 < s.size() && roman_map.at(s[i]) < roman_map.at(s[i + 1])) 
                result -= roman_map.at(s[i]);
            else 
                result += roman_map.at(s[i]);
        }

        return result;
    }
};
